### PR TITLE
Adding means to build the runtime docker images for arm32v7 and arm64v8

### LINKED ISSUE #3114: Build: Support Multi Arch
### What problem are you trying to solve?

Right now, if we want to build Linkerd to be able to work with another architecture for example like the ARM by running `bin/docker-build` inside Raspberry Pi, it will not build correctly(the resulting image is not runnable inside the Raspberry Pi) because of the existing image that is pulled for runtime image and also builder image is not supporting multi-arch[1]  therefore it will not run in another architecture.

image

https://github.com/linkerd/linkerd2/blob/f535c2d3d2f0da76bd62f3cbc1e39994a773ed21/bin/docker-build-base#L17-L23

https://github.com/linkerd/linkerd2/blob/f535c2d3d2f0da76bd62f3cbc1e39994a773ed21/bin/docker-build-go-deps#L17-L24

### How should the problem be solved?

It would be great if we can customize the target build architecture.
I am proposing to add a new environment key called `BUILD_ARCH` so that we can define in what architecture the resulting image should be based on.

The behavior should be like this:
if `BUILD_ARCH` NOT set: just do as current behavior do.

if `BUILD_ARCH` is set:
- Add suffix to the tag based on the BUILD_ARCH value ex: BUILD_ARCH=arm -> `gcr.io/linkerd-io/go-deps:4c8f4294.arm`, `gcr.io/linkerd-io/base:2019-02-19.01.arm`. With this addition of a suffix, the code will build the image instead of pulling it from the registry because it does not exist.
- Add `--platform=(BUILD_ARCH value)` when running the docker build[2], with this setting the docker will pull the base images with the defined platform.

Another improvement that I propose is to remove the hardcoded image string in the Dockerfile and move it as an argument, so we can put it in one single place that generates the tag(with or without suffix) and always add `--build-arg` when running the docker build.
```
docker build ... \
        --build-arg RUNTIME_IMAGE=repo:base_image_tag \
        BUILDER_IMAGE=repo:go_deps_tag
```

`base_image_tag` and `go_deps_tag` will generate the tag.
The output ex: `base_image_tag` = `base:2019-02-19.01`
The output ex: `go_deps_tag` = `go-deps:4c8f4294.arm`

### Any alternatives you've considered?

The solution that I do in order to build correctly for arm is changing the hardcoded tag(like add suffix) so it forces the code to build it instead of pulling it from registry, another solution is I just remove the code part that doing the docker pull, so straight to docker build without trying to pull. But all of this I need to perform inside my Raspberry Pi in order to docker detects my current architecture and use the correct images when building `base` and `go-deps`, another solution I can manually change the code part to use `--platform=arm` at the docker build. This is works because `debian:stretch-20190204-slim` and `golang:1.11.5` those images are supporting multi-arch.

With the solution I propose, now it is possible to do cross-build/cross-compile. I can just run `BUILD_ARCH=arm bin/docker-build` in the Mac and push the images to the registry and the raspberry pi can just pull it and run it properly.

### How would users interact with this feature?

```
BUILD_ARCH=arm bin/docker-build
BUILD_ARCH=arm64 bin/docker-build
```
This will result in the output images is based on the arm.
Supported `BUILD_ARCH` value 

https://github.com/docker/app/blob/d91c8816f56b9b2376f752a06dbd14394a78e4a3/vendor/github.com/containerd/containerd/platforms/database.go#L61

### Reference
[1] https://blog.docker.com/2017/11/multi-arch-all-the-things/
[2] https://docs.docker.com/engine/reference/commandline/build/

### Related issue
https://github.com/linkerd/linkerd2/issues/1165

### PR DESCRIPTION
**Pull request by jaerik**
Partly fixes #3114

Adding ability to build the runtime docker images for amd64, arm32v7, and arm64v8 regardless of host architecture and without the need of qemu (docker buildx). The solution take advantage the host architecture version of apt-get and dpkg to manage the target image's package needs.

### CODE REVIEW

**Comment by olix0r** [comment]
@jaerik This is really cool, thanks! Would you mind opening a PR against the linkerd2-proxy repo with the proxy's changes (i.e. replacing the patch file in this repo)? I want to test that out in the context of the proxy's CI process.

Thanks!

**Comment by jaerik** [comment]
> Would you mind opening a PR against the linkerd2-proxy repo with the proxy's changes (i.e. replacing the patch file in this repo)? I want to test that out in the context of the proxy's CI process.

@olix0r I have opened an PR [#497](https://github.com/linkerd/linkerd2-proxy/pull/497) against the linkerd2-proxy repo with the changes for the makefile.

**Comment by jaerik** [comment]
I don't see the connection between the failed test and the changes I made in the code. Have I done something wrong or is the test misleading?

**Comment by alpeb** [comment]
@jaerik it's ok it was just test flakiness

File: .proxy-init-version [hunk 1]
```diff
@@ -0,0 +1 @@
+v1.3.2
```
Comment (joakimr-axis) [inline .proxy-init-version:1]: Should probably be v1.3.3 now.

File: Dockerfile-proxy [hunk 2]
```diff
@@ -1,15 +1,21 @@
-ARG RUNTIME_IMAGE=debian:stretch-20190812-slim
+ARG ARCH=amd64
+ARG RUNTIME_IMAGE=$ARCH/debian:buster-20200422-slim
```
Comment (joakimr-axis) [inline Dockerfile-proxy:2]: Perhaps update to `buster-20200514-slim` which is what is now used for buster in other places in the codebase.

File: Dockerfile-proxy-init [hunk 3]
```diff
@@ -0,0 +1,61 @@
+ARG ARCH=amd64
+ARG GO_VERSION=1.12.9
+ARG VERSION=v1.3.2
```
Comment (joakimr-axis) [inline Dockerfile-proxy-init:3]: Should probably be v1.3.3 now.

File: Dockerfile-proxy-init [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ARG DEBIAN_VERSION=buster-20200422-slim
```
Comment (joakimr-axis) [inline Dockerfile-proxy-init:4]: See previous buster comment.

File: web/Dockerfile [hunk 5]
```diff
@@ -1,3 +1,8 @@
+ARG ARCH=amd64
+ARG RUNTIME_IMAGE=$ARCH/debian:buster-20200422-slim
```
Comment (joakimr-axis) [inline web/Dockerfile:2]: And here

**Comment by adleong** [comment]
@jaerik it looks like there's now a merge conflict on this PR that you'll need to resolve.

@olix0r now that associated linkerd2-proxy PR has merged, do you want to give this another look?

**Comment by aliariff** [comment]
Hi everyone,

I want to point out something, is that the approach of creating the Dockerfile for proxy-init & proxy and perform the compilation in this repo a correct approach? 
It seems like we duplicated the logic of build the image over here, as we already have our own repo for that in `linkerd2-proxy-init` & `linkerd2-proxy`, and that repos will have their own build & release process including supporting multiple architectures.
cc: @cpretzer @alpeb

**Comment by cpretzer** [comment]
hi @aliariff, you're right, the current plan is to keep the Dockerfile for `linkerd-proxy-init` and `linkerd-proxy` in their own repositories.

**Comment by jaerik** [comment]
> hi @aliariff, you're right, the current plan is to keep the Dockerfile for `linkerd-proxy-init` and `linkerd-proxy` in their own repositories.

Okay, I will move these parts to separate PR in the linkerd-proxy-init repository respective the linkerd-proxy repository.

**Comment by joakimr-axis** [comment]
> @olix0r changed the base branch from master to main 13 hours ago

Out of curiosity: What is this new `main` branch and how does it relate to `master`?

**Comment by olix0r** [comment]
main is replacing master. master will be deleted later this week to avoid confusion

**Comment by alpeb** [comment]
@jaerik Please let me know when this is again ready for review. I guess the extraction of the linkerd-proxy and linkerd-proxy-init is still pending, right?

**Comment by cpretzer** [comment]
@jaerik, thanks for your work on this, it set the foundation for a larger set of changes that we made in each of the repositories: [linkerd2-proxy](https://github.com/linkerd/linkerd2-proxy/pull/593), linkerd2-proxy-init [12](https://github.com/linkerd/linkerd2-proxy-init/pull/12/files) and [15](https://github.com/linkerd/linkerd2-proxy-init/pull/12/files), and [linkerd2](https://github.com/linkerd/linkerd2/pull/4765)

I'm going to close this one for now. If there is anything in the commits above that isn't covered, please let us know with a comment! 🙏

**Pull request closed without merging**